### ISSUE TITLE
colorRange is overwritten on subsequent hexbin layers

### ISSUE THREAD

**Issue by chriswilley** [Asymmetrik/leaflet-d3#32]
Thanks so much for writing this plugin collection, it's proving extremely helpful on a recent project.

For my project I am creating multiple hexbin layers, which can be toggled on and off by the user. They are added to a Leaflet LayerGroup when the application loads and then added to the map when the user clicks a button.

Prior to the current version (I started out using 1.4) I could set a colorRange for each layer and they would display correctly. As of v2.0 (and 2.0.1), the colorRange of the last layer added to the map is applied to all the layers. I have verified that each layer retains the correct options object, but they display using the most recent layer's colorRange.

I'm using this in an Angular2 application (thanks for creating the type definitions!) if that helps.

**Comment by reblace** [comment]
Thanks for the feedback! I'll look into this. I was just thinking about how multiple hexbin layers would intereact the other day. I haven't really ever tried it before. Do you have any concise code examples that reproduce the problem that I could use to test?

**Comment by chriswilley** [comment]
Probably, but I'll need to figure out how to extract the relevant code from the larger project.

**Comment by reblace** [comment]
Do you think you can reproduce it without angular 2? If so, you might try to create another standalone example file that reproduces the issue. Or just give a clear explanation of what you're doing and I can try to recreate it.

**Comment by chriswilley** [comment]
Good idea. Let's start with the clear explanation:

I'm iterating through an array of Javascript objects that describe the map layers using the following fields:

`[{
    displayName: '',
    date: '',
    latLngBounds: [],
    data: '',
    colorRange: []
},]`

"data" here is a geoJSON object that I pass to a function that strips out the lat/lon coordinates and provides them to the hexbinLayer.data() function:

`let options = {
    radius: 13,
    opacity: 1,
    colorRange: layer.colorRange // array of colors passed in from object above
};
let hexLayer = L.hexbinLayer(options);
let tempData = [];
for (let i = 0; i < layer.data.features.length; i++) {
    let coords = layer.data.features[i].geometry.coordinates;
    let lat = coords[0];
    let lon = coords[1];
    tempData.push([lat, lon]);
}
hexLayer.data(tempData);
L.layerGroup.addLayer(hexLayer);
`
All of this is happening inside ngOnInit() for the main Component of the application.

Then, when a user wants to view the hexbin layer, the button they click runs this function:

`
toggleLayer(lyr): void {
    let tempLayer = layerGroup.getLayer(lyr);
    if (this.baseMap.hasLayer(tempLayer) {
        this.baseMap.removeLayer(tempLayer);
    } else {
        this.baseMap.addLayer(tempLayer);
    }
}
`

Hope this helps. I'll see about reproducing in plain ol' Javascript.

**Comment by reblace** [comment]
I haven't read through your post, but figured I'd share this. I was able to get distinct color ranges working with multiple layers:

```js
<!DOCTYPE html>
<html>
<head>
	<title>Leaflet d3 HexBin Layer</title>

	<link rel="stylesheet" href="../../node_modules/leaflet/dist/leaflet.css" />

	<script src="../../node_modules/d3/build/d3.js" charset="utf-8"></script>
	<script src="../../node_modules/d3-hexbin/build/d3-hexbin.js" charset="utf-8"></script>
	<script src="../../node_modules/leaflet/dist/leaflet-src.js"></script>

	<script src="../../dist/leaflet-d3.js" charset="utf-8"></script>

<style>
	.hexbin-hexagon {
		stroke: #000;
		stroke-width: .5px;
	}
</style>

</head>

	<h2>Leaflet d3 Hexbin Example</h2>
	<p>Demonstrates using multiple layers</p>

	<!-- The map element -->
	<div id="map" style="width: 600px; height: 400px; border: 1px solid #ccc"></div>
	<br/>

	<input id="countInput" type="text" value="100"><button onclick="addLayer()">Add Layer</button>

	<script>

		var center = [ 38.991709, -76.886109 ];
		var layer = L.tileLayer('http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
			maxZoom: 18,
			attribution: '&copy; <a href="http://openstreetmap.org/copyright">OpenStreetMap</a> contributors'
		});

		var map = L.map('map', {
			layers: [ layer ],
			center: L.latLng(center[0], center[1]), zoom: 7
		});


		var latFn = function(c) { return d3.randomNormal(c[0], 0.3); }
		var longFn = function(c) { return d3.randomNormal(c[1], 0.3); }


		var layers = [
			{
			    center: [ 39.59, -76.39 ],
				colorRange: [ 'white', 'teal' ]
			},
			{
				center: [ 39.59, -77.39 ],
				colorRange: [ 'white', 'steelblue' ]
			},
			{
				center: [ 38.59, -77.39 ],
				colorRange: [ 'white', 'firebrick' ]
			},
			{
				center: [ 38.59, -76.39 ],
				colorRange: [ 'white', 'tomato' ]
			}
		];

		layers.forEach(function(e) {
			var hexLayer = L.hexbinLayer({ colorRange: e.colorRange }).addTo(map);
			var data = generateData(e.center);
			hexLayer.data(data);
		});

		function generateData(c) {
			var data = [];
			for(var i = 0; i < 500; i++) {
				data.push([ longFn(c)(), latFn(c)() ]);
			}
			return data;
		}

	</script>
</body>
</html>
```

**Comment by chriswilley** [comment]
Interesting. When I zoom in and then back out, the layers all appear to use the range ['white', 'tomato'].

**Comment by chriswilley** [comment]
Since I'm only noticing the problem when I view the layers close up, it appears that redrawing the layers is what causes the issue.

**Comment by reblace** [comment]
oooh ok, let me check that.

**Comment by reblace** [comment]
Check out the latest patch version (2.0.3), should fix this issue. Thanks for reporting the bug!

**Comment by chriswilley** [comment]
Fantastic! Works like a charm. Thanks for jumping on it so fast.

**Comment by reblace** [comment]
Fixed and verified in 2.0.3
